### PR TITLE
feat: add TorchGeo-style sampling utilities for segmentation

### LINKED ISSUE #761: Feature idea: shared TorchGeo-style sampling utilities for GeoAI vision tasks
### Discussed in https://github.com/opengeos/geoai/discussions/759

<div type='discussions-op-text'>

<sup>Originally posted by **Moritz-Langer** May 19, 2026</sup>
Hi,

thank you for this great repository. I found GeoAI while looking into DINOv3 as a segmentation backbone and really like the current implementation.

I may have missed something, but the DINOv3 segmentation example appears to rely on pre-generated image/mask chips via `export_geotiff_tiles` and then passes file lists into `DINOv3SegmentationDataset`.

For larger GeoTIFFs or workflows where chips should not be materialized beforehand, it could be useful to provide an optional TorchGeo-style sampling path. TorchGeo already supports geospatial samplers such as `RandomGeoSampler` and `GridGeoSampler` for querying `GeoDataset` by bounding boxes, which is useful for dynamic training and evaluation patches. 

A possible enhancement could be a small shared utility module, e.g. `geoai.utils.sampling` , that centralizes dataloader creation for different vision tasks.

This would not replace the current chip-based workflow, which is simple and useful, but provide an advanced alternative for users working directly with large rasters, dynamic sampling, or TorchGeo-based pipelines.

Potential benefits:

- avoid pre-exporting all chips
- support large rasters more naturally
- reuse sampling logic across classification, segmentation, and detection
- make DINOv3 fine-tuning easier to integrate into existing TorchGeo workflows

A possible drawback is that it might be necessary to provide different batch loading mechanisms, but this should be manageable.

This is just a first idea, but I think a centralized sampling utility could complement the current examples nicely.</div>

### PR DESCRIPTION
**Pull request by giswqs**
## Summary
- Add `geoai/utils/sampling.py` with TorchGeo-style dynamic patch sampling from large GeoTIFFs, so segmentation models can be trained without pre-exporting image chips.
- Add two example notebooks (`torchgeo_sampling_segmentation.ipynb`, `torchgeo_segmentation_simple.ipynb`) and unit tests for the new sampling module.
- Minor updates to `dinov3_finetune.py` plus codespell ignore list and bandit hook formatting tweaks.

## Test plan
- [x] `pre-commit run --all-files` passes
- [ ] `pytest tests/test_utils_sampling.py -v`
- [ ] `pytest tests/test_dinov3_finetune.py -v`
- [ ] Run the new example notebooks end to end

@Moritz-Langer

### CODE REVIEW

File: tests/test_utils_sampling.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                mask_path="tests/data/test_raster_single.tif",
+                output_dir=tmpdir,
+                chip_size=16,
+                train_length=2,
+                val_length=1,
+                batch_size=1,
+                num_epochs=1,
+                device="cpu",
+                verbose=False,
+            )
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/test_utils_sampling.py:313]: Done in 6cce3bdf — switched to `get_test_data_paths()`/`ensure_test_data_exists()` so the fixtures are generated on demand and resolved relative to the test file.

File: geoai/dinov3_finetune.py [hunk 2]
```diff
@@ -402,10 +402,29 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        target_key = "mask"
+        if not require_target and isinstance(batch, dict) and target_key not in batch:
+            target_key = None
+        x, y = geo_sample_to_tuple(
+            batch,
+            target_key=target_key,
+            num_channels=3,
+            normalize=True,
+            squeeze_target=True,
+        )
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/dinov3_finetune.py:429]: Done in 6cce3bdf — plumbed `num_channels` and `normalize_input` through `DINOv3Segmenter.__init__` (and `train_dinov3_segmentation`), and changed `geo_sample_to_tuple` to only divide by 255 when the input is integer dtype, so pre-normalized float tensors pass through unchanged.

File: geoai/utils/sampling.py [hunk 3]
```diff
@@ -0,0 +1,862 @@
...
+        raise ValueError("image must have shape (C, H, W) or (B, C, H, W)")
+
+    channel_dim = -3
+    channels = image.shape[channel_dim]
+    if channels == num_channels:
+        return image
+    if channels > num_channels:
+        index = [slice(None)] * image.ndim
+        index[channel_dim] = slice(0, num_channels)
+        return image[tuple(index)]
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/utils/sampling.py:420]: Done in 6cce3bdf — `_coerce_channels` now emits a `UserWarning` when bands are dropped, with a hint to subset the image explicitly to control band selection.

File: geoai/utils/sampling.py [hunk 4]
```diff
@@ -0,0 +1,862 @@
...
+                loss = loss_fn(logits, masks)
+                if training:
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+
+            total_loss += loss.item()
+            total_iou += _segmentation_mean_iou(logits.detach(), masks, num_classes)
+            count += 1
+
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/utils/sampling.py:735]: Done in 6cce3bdf — `run_epoch` now raises a descriptive `ValueError` when the dataloader yields no batches instead of falling through to a `ZeroDivisionError`.
